pre_commit_hooks/ros_include_guard.py

```diff
@@ -0,0 +1,165 @@
+            self.tokens[2] = macro_name
+            self.expected = "".join(self.tokens)
+        else:
+            self.expected = f"{self.directive} {macro_name}"
+
+
+class ClosingDirective(DirectiveBase):
+    def prepare(self, macro_name):
+        self.tokens = split_space_boundary(self.text)
+        if 5 <= len(self.tokens):
+            self.tokens[4] = macro_name
+            self.expected = "".join(self.tokens)
+        else:
+            self.expected = f"{self.directive}  // {macro_name}"
+
+
+class IncludeGuard:
+    def __init__(self):
+        self.ifndef = OpeningDirective("#ifndef")
+        self.define = OpeningDirective("#define")
+        self.endif = ClosingDirective("#endif")
+        self.pragma = False
+
+    def items(self):
+        yield self.ifndef
+        yield self.define
+        yield self.endif
+
+    def is_none(self):
+        return any(item.is_none() for item in self.items())
+
+    def mismatch(self):
+        return any(item.mismatch() for item in self.items())
+
+    def overwrite(self, lines):
+        return any(item.overwrite(lines) for item in self.items())
+
+    def prepare(self, macro_name):
+        for item in self.items():
+            item.prepare(macro_name)
+
+
+def split_space_boundary(text, delimiters={" "}):
+
+    result = []
+    prev = None
+    for char in text:
+        current = char in delimiters
+        if current != prev:
+            result.append(char)
+        else:
+            result[-1] += char
+        prev = current
+    return result
+
+
+def get_include_guard_info(lines):
+
+    guard = IncludeGuard()
+    for line, text in enumerate(lines):
+        if text.startswith("#pragma once"):
+            guard.pragma = True
+        if text.startswith("#ifndef") and guard.ifndef.is_none():
+            guard.ifndef.update(line, text)
+        if text.startswith("#define") and guard.define.is_none():
+            guard.define.update(line, text)
+        if text.startswith("#endif"):
+            guard.endif.update(line, text)
+    return guard
+
+
+def get_parts_after(parts, targets):
+    result = []
+    for part in reversed(parts):
+        if part in targets:
+            break
+        result.append(part)
+    return reversed(result)
+
+
+def get_include_guard_macro_name(filepath):
+    targets = {"include", "src", "test"}
+    parts = get_parts_after(filepath.parts, targets)
+    return "__".join(parts).replace(".", "_").upper() + "_"
+
+
+def main(argv=None):
+
+    parser = argparse.ArgumentParser()
+    parser.add_argument("filenames", nargs="*", help="Filenames to fix")
+    args = parser.parse_args(argv)
+
+    return_code = 0
+    for filename in args.filenames:
+        # Search the include guard lines.
+        filepath = pathlib.Path(filename)
+        lines = filepath.read_text().split("\n")
+        guard = get_include_guard_info(lines)
+        # Skip if the file using pragma once.
+        if guard.pragma:
+            continue
```

Comment:
Why you continue here is that it's difficult to consider the cases that have both `#define` and `#pragma once`?

Reply:
I didn't expect that case. I'll add error If there are both.

Reply:
Thank you, but please note that it's sometimes done intendedly.
https://stackoverflow.com/questions/50090382/why-using-both-pragma-once-and-include-guard

Reply:
I think we can ignore this case for now.

Reply:
https://github.com/tier4/pre-commit-hooks-ros/pull/32/commits/04cf4719e85c87a2b7b5a52a6841fd5be8945fea
- fix macro name even if there are pragma once.
- suppress error of no include guard if there are pragma once.